src/rpc/notifications.cpp

```diff
@@ -0,0 +1,381 @@
+        throw std::runtime_error(
+            "getregisterednotifications <uuid>\n"
+            "\nReturns the currently registered RPC notification types for the given uuid.\n"
+            "\nArguments:\n"
+            "1. \"uuid\"         (string, required) The client uuid\n"
+            "\nResult:\n"
+            "\"[\"\n"
+            "\"  \"<signal>\"             (string) The registered signal\n"
+            "\"  ,...\n"
+            "\"]\"\n"
+            "\nExamples:\n"
+            "\nCreate a transaction\n" +
+            HelpExampleCli("getregisterednotifications", "") +
+            "Get the registered notification types\n" + HelpExampleRpc("getregisterednotifications", ""));
+
+    RPCTypeCheck(request.params, boost::assign::list_of(UniValue::VSTR), true);
+    NotificationQueue* clientQueue = getQueue(request.params[0].get_str(), false);
+
+    UniValue result = UniValue(UniValue::VARR);
+    BOOST_FOREACH (NotificationType type, clientQueue->m_registered_notification_types)
+        result.push_back(clientQueue->typeToString(type));
+
+    return result;
+}
+
+UniValue pollnotifications(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 2)
+        throw std::runtime_error(
+            "pollnotifications <uuid> <timeout>\n"
+            "\nLong poll function to get all available notifications for a given uuid (see setregisterednotifications for how to register for notifications).\n"
+            "The RPC thread will idle for the via <timeout> defined amount of seconds and/or will immediately response if new notifications are available\n"
+            "Arguments:\n"
+            "1. \"uuid\"         (string, required) The client uuid\n"
+            "2. \"timeout\"      (numeric, optional) The timeout \n"
+            "\nResult:\n"
+            "\"[ notification, ... ]\"             (object) The notification object\n"
+            "\nExamples:\n"
+            "\nPoll notifications for client a8098c1a...\n" +
+            HelpExampleCli("pollnotifications", "\"a8098c1a-f86e-11da-bd1a-00112444be1e\" 500") +
+            "Long poll notification (max. 500 seconds)\n" + HelpExampleRpc("pollnotifications", "\"a8098c1a-f86e-11da-bd1a-00112444be1e\" 500"));
+
+    RPCTypeCheck(request.params, boost::assign::list_of(UniValue::VSTR)(UniValue::VNUM), false);
+    NotificationQueue* clientQueue = getQueue(request.params[0].get_str(), false);
+
+    int64_t timeOut = DEFAULT_POLL_TIMEOUT;
+    if (request.params.size() == 2)
+        timeOut = request.params[1].get_int64();
+
+    int64_t startTime = GetTime();
+
+    UniValue result = UniValue(UniValue::VARR);
+    // allow long polling
+    while (!ShutdownRequested()) {
+        if (clientQueue->elementsAvailable()) {
+            clientQueue->dequeueElements(result);
+            break;
+        }
+        if (startTime + timeOut + (500 / 1000.0) < GetTime())
+            break;
+        MilliSleep(500);
```

Comment:
Should probably use a CV and a way for Interrupt to interrupt it instead of calling ShutdownRequested in a 500ms loop.